spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+    Monitoring monitoring;
+
+    try {
+      HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+      JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+      GoogleCredential credential = loadCredential(transport, jsonFactory,
+                                                   credentialsPath);
+      Monitoring.Builder builder = new Monitoring.Builder(
+            transport, jsonFactory, credential);
+
+        builder.setApplicationName("Spectator");
+      monitoring = builder.build();
+    } catch (IOException|java.security.GeneralSecurityException e) {
+      throw new IllegalStateException(e);
+    }
+
+    log.info("Created StackdriverMonitoring client for project '{}'",
+             projectName);
+    return new StackdriverWriter(monitoring, customTypeNamespace,
+                                 projectName,
+                                 applicationName, applicationVersion,
+                                 measurementFilter);
+  }
+
+  /**
+   * Helper function for the constructor that reads our credentials for
+   * talking to Stackdriver.
+   */
+  private static GoogleCredential loadCredential(
+      HttpTransport transport, JsonFactory factory, String credentialsPath)
+      throws IOException {
+    GoogleCredential credential;
+    if (!credentialsPath.isEmpty()) {
+      FileInputStream stream = new FileInputStream(credentialsPath);
+      credential = GoogleCredential.fromStream(stream, transport, factory)
+                   .createScoped(
+                         Collections.singleton(MonitoringScopes.MONITORING));
+      log.info("Loaded credentials from from {}", credentialsPath);
+    } else {
+      log.info("spectator.stackdriver.monitoring.enabled without"
+               + " spectator.stackdriver.credentialsPath. "
+               + " Using default application credentials.");
+      credential = GoogleCredential.getApplicationDefault();
+    }
+    return credential;
+  }
+
+  /**
+   * Constructs a writer instance.
+   */
+  public StackdriverWriter(
+      Monitoring monitoring,
+      String customTypeNamespace,
+      String projectName, String applicationName, String applicationVersion,
+      MeasurementFilter measurementFilter) {
+    if (projectName.isEmpty()) {
+      throw new IllegalStateException(
+          "You need to configure a stackdriver.projectName.");
+    }
+    this.service = monitoring;
+    this.customTypePrefix = String.format("custom.googleapis.com/%s/",
+                                          customTypeNamespace);
+    this.projectResourceName = "projects/" + projectName;
+    this.applicationName = applicationName;
+    this.measurementFilter = measurementFilter;
+
+    UUID uuid = UUID.randomUUID();
+    byte[] uuidBytes = new byte[16];
+    addLong(uuidBytes, 0, uuid.getLeastSignificantBits());
+    addLong(uuidBytes, 8, uuid.getMostSignificantBits());
+    setInstanceId(java.util.Base64.getEncoder().encodeToString(uuidBytes));
```

Comment:
What is the usage of instance id? For example this will change if the process is restarted. I was expecting something more like a host name or ec2 instance id that can be looked up in other systems.